lib/Sema/DerivedConformances.cpp

```diff
@@ -380,3 +389,215 @@ bool DerivedConformance::checkAndDiagnoseDisallowedContext(
 
   return false;
 }
+
+/// Build a type-checked integer literal.
+static IntegerLiteralExpr *buildIntegerLiteral(ASTContext &C, unsigned index) {
+  Type intType = C.getIntDecl()->getDeclaredType();
+
+  auto literal = IntegerLiteralExpr::createFromUnsigned(C, index);
+  literal->setType(intType);
+  literal->setBuiltinInitializer(C.getIntBuiltinInitDecl(C.getIntDecl()));
+
+  return literal;
+}
+
+/// Create AST statements which convert from an enum to an Int with a switch.
+/// \p stmts The generated statements are appended to this vector.
+/// \p parentDC Either an extension or the enum itself.
+/// \p enumDecl The enum declaration.
+/// \p enumVarDecl The enum input variable.
+/// \p funcDecl The parent function.
+/// \p indexName The name of the output variable.
+/// \return A DeclRefExpr of the output variable (of type Int).
+DeclRefExpr *swift::convertEnumToIndex(SmallVectorImpl<ASTNode> &stmts,
```

Comment:
Can we limit the scope of this API promotion to DerivedConformance?

Reply:
see above